src/sentry/preprod/models.py

```diff
@@ -0,0 +1,149 @@
+        AAB = 3
+        AAB_ZIPPED = 4
+        APK = 5
+        APK_ZIPPED = 6
+
+        @classmethod
+        def as_choices(cls):
+            return (
+                (cls.XCARCHIVE, "xcarchive"),
+                (cls.XCFRAMEWORK, "xcframework"),
+                (cls.IPA, "ipa"),
+                (cls.AAB, "aab"),
+                (cls.AAB_ZIPPED, "aab_zipped"),
+                (cls.APK, "apk"),
+                (cls.APK_ZIPPED, "apk_zipped"),
+            )
+
+    class ErrorCode:
+        UNKNOWN = 0
+        """The error code is unknown. Try to use a descriptive error code if possible."""
+        UPLOAD_TIMEOUT = 1
+        """The upload timed out."""
+        ARTIFACT_PROCESSING_TIMEOUT = 2
+        """The artifact processing timed out."""
+        ARTIFACT_PROCESSING_ERROR = 3
+        """The artifact processing failed."""
+
+        @classmethod
+        def as_choices(cls):
+            return (
+                (cls.UNKNOWN, "unknown"),
+                (cls.UPLOAD_TIMEOUT, "upload_timeout"),
+                (cls.ARTIFACT_PROCESSING_TIMEOUT, "artifact_processing_timeout"),
+                (cls.ARTIFACT_PROCESSING_ERROR, "artifact_processing_error"),
+            )
+
+    # Having a FK to both Org/Project is unnecessary
+    organization_id = BoundedBigIntegerField(db_index=True)
+    project = FlexibleForeignKey("sentry.Project")
+
+    # Nullable in case the file upload fails
+    file = FlexibleForeignKey("sentry.File", null=True)
+
+    # The date the artifact was built. E.g. an artifact could be built on 05/21/2025,
+    # but the user uploaded it on 05/22/2025.
+    date_built = models.DateTimeField(null=True)
+
+    build_configuration = models.ForeignKey("sentry.PreprodBuildConfiguration", null=True)
+
+    state = BoundedPositiveIntegerField(
+        default=ArtifactState.UPLOADING, choices=ArtifactState.as_choices()
+    )
+
+    artifact_type = BoundedPositiveIntegerField(choices=ArtifactType.as_choices(), null=True)
+
+    error_code = BoundedPositiveIntegerField(choices=ErrorCode.as_choices(), null=True)
+    error_message = models.TextField(null=True)
+
+    # E.g. 1.2.300
+    build_version = models.CharField(max_length=255, null=True)
+    # E.g. 9999
+    build_number = models.IntegerField(null=True)
+    # Some artifacts are embedded with a UUID, e.g. iOS Mach-O binaries
+    build_uuid = models.UUIDField(null=True)
+
+    misc = models.Field[dict[str, Any], dict[str, Any]] = JSONField(null=True)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_preprodartifact"
+
+
+class PreprodBuildConfiguration(DefaultFieldsModel):
+    organization_id = BoundedBigIntegerField(db_index=True)
+    project = FlexibleForeignKey("sentry.Project")
+    name = models.CharField(max_length=255)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_preprodbuildconfiguration"
+        unique_together = ("project", "name")
+
+
+class PreprodArtifactSizeMetrics(DefaultFieldsModel):
```

Comment:
One thought I had while writing this was having an additional `SizeMetricsType` field for `main_app`, `watch_app`, `dynamic_feature`, which might solve the subproduct issue mentioned in the doc. Each `PreprodArtifact` could then have many `PreprodArtifactSizeMetrics`. This would still be more ergonomic than before because rarely do you need to query the size metrics unless you are doing something immediately with the results.